tech-matters/serverless
# Create customizable function to handle operating hours

Jira: https://bugs.benetech.org/browse/CHI-532

This PR adds customizable `operatingHours` function. The customization can be done through a json file under the assets folder. We can then have all of the operating hours bundled within our assets, refer to the appropriate one via envvar and update them with a deploy, but that requires no code changes. 

To test this PR:
- Add `OPERATING_INFO_KEY=za` to your local `.env` file.
- For ease, you can set `const TokenValidator = (f: any) => f;` or remove the `TokenValidator` wrapper (that way you don't need to provide a token). If you do this you also must remove the `: ServerlessFunctionSignature` type annotation in `export const handler: ServerlessFunctionSignature`.
- Start the local development server with `npm run start`.
- Hit `operatingHours` and see response (`curl -X GET 'http://localhost:3000/operatingHours'`).
- You can check the current time in ZA [here](https://www.google.com/search?q=africa%2Fjohannesburg+timezone&sxsrf=ALeKk03m0x_gH9vAgYeWMGwmgddzIc-s-A%3A1617825791066&ei=_w9uYJvDA8XR5OUPmeiH4AM&oq=africa%2Fjohannesburg+timezo&gs_lcp=Cgdnd3Mtd2l6EAMYADICCAAyBggAEBYQHjoHCAAQRxCwAzoECAAQHjoJCAAQyQMQFhAeUNgZWKsiYOEoaAFwAngAgAGFAYgB1AaSAQM1LjOYAQCgAQGqAQdnd3Mtd2l6yAEHwAEB&sclient=gws-wiz) to see if the result is correct while avoid changing the timezone in the json file. You can change the open/close values in the corresponding day to test behavior locally. Also, you can add the current date with `MM/DD/YYYY` format to the holidays object to test it.

~WIP:~
- ~Add tests (in case this approach is approved).~
- ~Separate the `operatingHours` by channel. This can be done nesting the OperatingInfo -> operatingHours under channel key string (i.e. { "facebook": { "1": { open: xx, "close": yy }, ... },  "webchat": { ... }  }). This assumes that `holidays` and `timezone` will be shared across channels (what I believe, is correct). Please suggest if there's any other approach that makes more sense to you!~
